aws_lambda_builders/workflows/rust_cargo/actions.py

```diff
@@ -0,0 +1,228 @@
+from aws_lambda_builders.workflow import BuildMode
+from aws_lambda_builders.actions import ActionFailedError, BaseAction, Purpose
+
+
+class BuilderError(Exception):
+    MESSAGE = "Builder Failed: {message}"
+
+    def __init__(self, **kwargs):
+        Exception.__init__(self, self.MESSAGE.format(**kwargs))
+
+
+class OSUtils(object):
+    """
+    Wrapper around file system functions, to make it easy to
+    unit test actions in memory
+    """
+
+    def popen(self, command, stdout=None, stderr=None, env=None, cwd=None):
+        return subprocess.Popen(command, stdout=stdout, stderr=stderr, env=env, cwd=cwd)
+
+    def copyfile(self, source, destination):
+        shutil.copyfile(source, destination)
+
+    def makedirs(self, path):
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+
+def parse_handler(handler):
+    """
+    Parse package and binary from handler name
+
+    If the handler contains a period, assume `package.bin_name`
+    otherwise assume common case where bin_name is the same as the package
+    """
+    spec = handler.split(".", 1)
+    if len(spec) == 1:
+        return (spec[0], spec[0])
+    else:
+        return (spec[0], spec[1])
+
+
+class BuildAction(BaseAction):
+    NAME = "CargoBuild"
+    DESCRIPTION = "Building the project using Cargo"
+    PURPOSE = Purpose.COMPILE_SOURCE
+
+    def __init__(self, source_dir, handler, binaries, platform, mode, osutils=OSUtils()):
+        """
+        Build the a rust executable
+
+        :type source_dir: str
+        :param source_dir:
+            Path to a folder containing the source code
+
+        :type handler: str
+        :param handler:
+            Handler name in `package.bin_name` or `bin_name` format
+
+        :type binaries: dict
+        :param binaries:
+            Resolved path dependencies
+
+        :type platform: string
+        :param platform:
+            Platform builder is being run on
+
+        :type mode: str
+        :param mode:
+            Mode the build should produce
+
+        :type osutils: object
+        :param osutils:
+            Optional, External IO utils
+        """
+        self.source_dir = source_dir
+        self.handler = handler
+        self.mode = mode
+        self.binaries = binaries
+        self.platform = platform
+        self.osutils = osutils
+
+    def cargo_metadata(self):
+        p = self.osutils.popen(
+            ["cargo", "metadata", "--no-deps", "--format-version=1"],
+            stderr=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            env=os.environ.copy(),
+            cwd=self.source_dir,
+        )
+        out, err = p.communicate()
+        if p.returncode != 0:
+            raise BuilderError(message=err.decode("utf8").strip())
+        return json.loads(out)
+
+    def build_command(self, package):
+        cmd = [self.binaries["cargo"].binary_path, "build", "-p", package, "--target", "x86_64-unknown-linux-musl"]
+        if self.mode != BuildMode.DEBUG:
```

Comment:
> Does Rust build a debuggable artifact by default unless --release is added?

Correct. Rust builds debug artifacts by default. The default assumes you are building something you will run locally then asks you to pass --release for a deployment optimized artifact.

Sam is the opposite you assume deployment artifact by default and debug on demand

 Inverting would largely depend on Sam or some higher level construct consitently passing debug mode workflow rather than something like None.

Reply:
Let me know if you feel strongly about this. I'd like to try to make the implementation follow sams semantics, default to deploy unless opting into debug.